templates/common/on-prem/files/NetworkManager-resolv-prepender.yaml

```diff
@@ -20,89 +20,124 @@ contents:
+    }
 
-       if [[ -z "$STATIC_HOSTNAME" || "$STATIC_HOSTNAME" == "localhost.localdomain" ]] ; then
- 
-          # run with systemd-run to avoid selinux problems
-          systemd-run --property=Type=oneshot --unit resolve-prepender-hostnamectl -Pq \
-              hostnamectl set-hostname --static --transient $DHCP6_FQDN_FQDN
-       fi
-    fi
-    
-    case "$STATUS" in
-        up|dhcp4-change|dhcp6-change)
-        >&2 echo "NM resolv-prepender triggered by ${1} ${2}."
+    function resolv_prepender {
+        # If $DHCP6_FQDN_FQDN is not empty and is not localhost.localdomain and static hostname was not already set
+        if [[ -n "$DHCP6_FQDN_FQDN" && "$DHCP6_FQDN_FQDN" != "localhost.localdomain" && "$DHCP6_FQDN_FQDN" =~ "." ]] ; then
+           STATIC_HOSTNAME="$(test ! -e /etc/hostname && echo -n || cat /etc/hostname | xargs)"
+
+           if [[ -z "$STATIC_HOSTNAME" || "$STATIC_HOSTNAME" == "localhost.localdomain" ]] ; then
+              # run with systemd-run to avoid selinux problems
+              systemd-run --property=Type=oneshot --unit resolve-prepender-hostnamectl -Pq \
+                  hostnamectl set-hostname --static --transient $DHCP6_FQDN_FQDN
+           fi
+        fi
+
+        case "$STATUS" in
+            up|dhcp4-change|dhcp6-change)
+            >&2 echo "NM resolv-prepender triggered by ${IFACE} ${STATUS}."
 
-        # In DHCP connections, the resolv.conf content may be late, thus we wait for nameservers
-        timeout 20s /bin/bash <<EOF
-            >&2 echo  "NM resolv-prepender: Checking for nameservers in /var/run/NetworkManager/resolv.conf"
+            # In DHCP connections, the resolv.conf content may be late, thus we wait for nameservers
             while ! grep nameserver /var/run/NetworkManager/resolv.conf; do
                 >&2 echo  "NM resolv-prepender: NM resolv.conf still empty of nameserver"
                 sleep 0.5
             done
-    EOF
-        # Ensure resolv.conf exists and contains nameservers before we try to run podman
-        if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
-            cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
-        fi
 
-        # Pull container image outside of the timeout to workaround possible
-        # image corruption
-        # https://github.com/containers/podman/issues/14003
-        /usr/bin/podman pull --authfile /var/lib/kubelet/config.json {{ .Images.baremetalRuntimeCfgImage }}
+            # This function is subject to a timeout. The timeout may interrupt podman run's image pull if the image is
+            # not available and cause image corruption [1]. Image should be available or being pulled before we reach
+            # this point. Wait until its available.
+            # [1] https://github.com/containers/podman/issues/14003
+            until /usr/bin/podman image exists {{ .Images.baremetalRuntimeCfgImage }}; do
```

Comment:
This can't be done before we populate resolv.conf on line 62. If I'm reading this right, the script will never succeed as written because the image pull won't work until resolv.conf is populated, and this check prevents us from ever populating resolv.conf.

I think we need to factor the resolv.conf creation out and run it first, and then we can start the image pull in the background and only wait to prepend the local IP.

Reply:
`podman image exists` is a local check I think.

But you are right, I think 
```
# Ensure resolv.conf exists and contains nameservers before we try to run podman
if [[ ! -e /etc/resolv.conf ]] || ! grep -q nameserver /etc/resolv.conf; then
    cp /var/run/NetworkManager/resolv.conf /etc/resolv.conf
fi
```
should be moved into `pull_baremetal_runtime_cfg_image`